docs/web/howto/client.rst

```diff
@@ -331,74 +331,85 @@ This function returns a ``Deferred`` that fires with the body after the request
+The following Python program should produce a verification error, when run as ``python example.py https://wrong.host.badssl.com/``.
 
 .. code-block:: python
 
-    
-    from twisted.python.log import err
-    from twisted.web.client import Agent
-    from twisted.internet import reactor
-    from twisted.internet.ssl import optionsForClientTLS
-
-    def display(response):
-        print("Received response")
-        print(response)
-    
-    def main():
-        contextFactory = optionsForClientTLS(u"https://example.com/")
-        agent = Agent(reactor, contextFactory)
-        d = agent.request("GET", "https://example.com/")
-        d.addCallbacks(display, err)
-        d.addCallback(lambda ignored: reactor.stop())
-        reactor.run()
-    
-    if __name__ == "__main__":
-        main()
+    import sys
 
+    from twisted.internet.task import react
 
+    from twisted.web.client import Agent, ResponseFailed
 
-    
-
+    @react
+    def main(reactor):
+        agent = Agent(reactor)
+        requested = agent.request(b"GET", sys.argv[1])
+        def gotResponse(response):
+            print(response.code)
+        def noResponse(failure):
+            failure.trap(ResponseFailed)
+            print(failure.value.reasons[0].getTraceback())
+        return requested.addCallbacks(gotResponse, noResponse)
 
+This is because the certificate returned by ``wrong.host.badssl.com`` is actually for ``badssl.com``, and the hostnames don't match.
+In our pretend scenario, we want to construct an ``Agent`` that validates HTTPS certificates normally, _except_ for this one host, where it ignores the hostname mismatch and supplies the correct hostname manually, to work around this problem.
+In order to do that, we will supply our own policy that creates a per - netloc–i.e. per host+port combination – TLS configuration, like so:
 
-For more fine-grained over the TLS configuration, check out the documentation for the :api:`twisted.internet.ssl <ssl>` module and :doc:`Using SSL in Twisted <../../core/howto/ssl>`.
+.. code-block:: python
 
+    import sys
 
+    from zope.interface import implementer
 
+    from twisted.internet.task import react
+    from twisted.internet.ssl import optionsForClientTLS
 
+    from twisted.web.iweb import IPolicyForHTTPS
+    from twisted.web.client import Agent, ResponseFailed, BrowserLikePolicyForHTTPS
+
+    @implementer(IPolicyForHTTPS)
+    class OneHostnameWorkaroundPolicy(object):
+        def __init__(self):
+            self._normalPolicy = BrowserLikePolicyForHTTPS()
+        def creatorForNetloc(self, hostname, port):
+            if hostname == b"wrong.host.badssl.com":
+                return optionsForClientTLS(u"badssl.com")
+            else:
+                return self._normalPolicy.creatorForNetloc(hostname, port)
+
+    @react
+    def main(reactor):
+        agent = Agent(reactor, OneHostnameWorkaroundPolicy())
+        requested = agent.request(b"GET", sys.argv[1])
```

Comment:
OK.  Thanks for pointing this out.  Lucky for us, implicitly, the type of `IAgent.request`'s second argument is actually "ASCII-only bytes", because IRIs officially aren't 8-bit clean.